src/inmanta/deploy/scheduler.py

```diff
@@ -864,98 +864,93 @@ async def _new_version(
+                    self._state.resource_state[resource].blocked is BlockedStatus.TRANSIENT
+                    # it shouldn't be any longer
+                    and not self._state.should_skip_for_dependencies(resource)
+                ):
+                    self._state.resource_state[resource].blocked = BlockedStatus.NO
+
+            # Update set of in-progress deploys that became unmanaged
+            self._deploying_unmanaged.update(self._deploying_latest & (new | deleted))
+            # Update set of in-progress non-stale deploys by trimming resources with new state
+            self._deploying_latest.difference_update(intent_changes.keys(), transitive_blocked)
+
+            # Remove timers for resources that are:
+            #    - in the dirty set (because they will be picked up by the scheduler eventually)
+            #    - blocked: must not be deployed
+            #    - deleted from the model
+            self._timer_manager.stop_timers(self._state.dirty | became_undefined | transitive_blocked)
+            self._timer_manager.remove_timers(deleted)
+            # Install timers for initial up-to-date resources. They are up-to-date now,
+            # but we want to make sure we periodically repair them.
+            self._timer_manager.update_timers(
+                up_to_date_resources | (transitive_unblocked - self._state.dirty), are_compliant=True
+            )
 
-                await self.state_update_manager.update_resource_intent(
-                    self.environment,
-                    intent={
-                        rid: (self._state.resource_state[rid], self._state.resources[rid])
-                        for rid in resources_with_updated_blocked_state
-                    },
-                    update_blocked_state=True,
-                    connection=con,
-                )
+            # ensure deploy for ALL dirty resources, not just the new ones
+            self._work.deploy_with_context(
+                self._state.dirty,
+                reason=reason,
+                priority=TaskPriority.NEW_VERSION_DEPLOY,
+                deploying=self._deploying_latest,
+                added_requires=added_requires,
+                dropped_requires=dropped_requires,
+            )
+            for resource in deleted:
+                self._state.drop(resource)  # Removes from the dirty set
+            for resource in deleted | became_undefined | transitive_blocked:
+                self._work.delete_resource(resource)
+
+            # Updating the blocked state should be done under the scheduler lock, because this state is written
+            # by both the deploy and the new version code path.
+            resources_with_updated_blocked_state: Set[ResourceIdStr] = (
+                became_undefined | transitive_blocked | transitive_unblocked
+            )
 
+        async with self.state_update_manager.get_connection(connection=connection) as con, con.transaction():
```

Comment:
So the intent lock keeps this neatly sequential.

Reply:
Well, not quite, because who knows what the IO loop may do in the meantime

Reply:
Oh, intent lock... Yes, that's the idea.